lib/install.js

```diff
@@ -136,71 +168,99 @@ async function install (fs, gyp, argv) {
+          )
+        } catch (err) {
+          // something went wrong downloading the tarball?
+          if (err.code === 'ENOTFOUND') {
+            throw new Error('This is most likely not a problem with node-gyp or the package itself and\n' +
+              'is related to network connectivity. In most cases you are behind a proxy or have bad \n' +
+              'network settings.')
+          }
+          throw err
         }
-        throw err
       }
-    }
 
-    // invoked after the tarball has finished being extracted
-    if (extractCount === 0) {
-      throw new Error('There was a fatal problem while downloading/extracting the tarball')
-    }
+      // invoked after the tarball has finished being extracted
+      if (extractErrors || extractCount === 0) {
+        throw new Error('There was a fatal problem while downloading/extracting the tarball')
+      }
+
+      log.verbose('tarball', 'done parsing tarball')
+
+      const installVersionPath = path.resolve(tarExtractDir, 'installVersion')
+      await Promise.all([
+        // need to download node.lib
+        ...(win ? downloadNodeLib() : []),
+        // write the "installVersion" file
+        fs.promises.writeFile(installVersionPath, gyp.package.installVersion + '\n'),
+        // Only download SHASUMS.txt if we downloaded something in need of SHA verification
+        ...(!tarPath || win ? [downloadShasums()] : [])
+      ])
+
+      log.verbose('download contents checksum', JSON.stringify(contentShasums))
+      // check content shasums
+      for (const k in contentShasums) {
+        log.verbose('validating download checksum for ' + k, '(%s == %s)', contentShasums[k], expectShasums[k])
+        if (contentShasums[k] !== expectShasums[k]) {
+          throw new Error(k + ' local checksum ' + contentShasums[k] + ' not match remote ' + expectShasums[k])
+        }
+      }
 
-    log.verbose('tarball', 'done parsing tarball')
-
-    const installVersionPath = path.resolve(devDir, 'installVersion')
-    await Promise.all([
-      // need to download node.lib
-      ...(win ? downloadNodeLib() : []),
-      // write the "installVersion" file
-      fs.promises.writeFile(installVersionPath, gyp.package.installVersion + '\n'),
-      // Only download SHASUMS.txt if we downloaded something in need of SHA verification
-      ...(!tarPath || win ? [downloadShasums()] : [])
-    ])
-
-    log.verbose('download contents checksum', JSON.stringify(contentShasums))
-    // check content shasums
-    for (const k in contentShasums) {
-      log.verbose('validating download checksum for ' + k, '(%s == %s)', contentShasums[k], expectShasums[k])
-      if (contentShasums[k] !== expectShasums[k]) {
-        throw new Error(k + ' local checksum ' + contentShasums[k] + ' not match remote ' + expectShasums[k])
+      // copy over the files from the temp tarball extract directory to devDir
+      if (tarExtractDir !== devDir) {
+        await copyDirectory(tarExtractDir, devDir)
```

Comment:
> Is there a reason we wouldn't want to `rename` / `move` here instead of copy?

Since there are parallel installs happening, there's no way to atomically rename to `devDir`. That directory is created (if it doesn't already exist) [at the start of the install process to confirm permissions](https://github.com/nodejs/node-gyp/blob/6f3c2d3c6c0de0dbf8c7245f34c2e0b3eea53812/lib/install.js#L106), so a direct rename of `tarExtractDir` to `devDir` would fail. Any rename process would need to first rename `devDir`, which opens a race condition  where `devDir` disappears momentarily during an ongoing build, causing a failure. That was [my original approach](https://github.com/dsanders11/node-gyp/commit/b21e845dd3217a0c4e5461af36962a275736f693#diff-a021b62b61739461595e009ed3d8524495144ddba458d99f1eba19e2a2be7b74R247-R255) for fixing this issue, but I switched to the current implementation to close that race condition.

> This could still race condition if it's not atomic?

The `copyFile` approach stays similar to the existing implementation, with `node-tar` overwriting the existing files when there are parallel installs going on. The main difference is that on Windows `node-tar` always unlinks the destination file path before writing to it to avoid certain conditions (same path multiple times in an archive) which don't exist for this usage.

I've stress tested this change (on CI and locally with a real package) with 35 parallel installs and didn't trip over any race conditions in practice.